src/Microsoft.Azure.SignalR/HubHost/ServiceConnection.cs

```diff
@@ -102,11 +102,18 @@ private async Task ProcessOutgoingMessagesAsync(ServiceConnectionContext connect
             }
             catch (Exception ex)
             {
+                // The exception means applicaion fail to process input anymore
+                // Turn down application output to avoid new messages and notify client to disconnect
                 Log.SendLoopStopped(_logger, connection.ConnectionId, ex);
+                connection.Application.Output.CancelPendingFlush();
+                connection.Application.Output.Complete();
+                await WriteAsync(new CloseConnectionMessage(connection.ConnectionId, ex.Message));
             }
             finally
             {
+                // Complete Input to let WaitOnApplicationTask() clean transport.
                 connection.Application.Input.Complete();
+                RemoveClientConnection(connection.ConnectionId);
```

Comment:
call `PerformDisconnectAsyncCore` instead?

Reply:
Will it make ApplicationTask and current Task waiting for each other?

Reply:
After we complete application Input and Output, the only thing in PerformDisconnectAsyncCore() is to RemoveConnection, so I directly call it in the end.

Reply:
I forget about in normal case, it will also do RemoveClientConnection() which will break normal disconnect. Will fix this.